File: algorithms/leetcode/longest-substring-without-repeating-characters/longest_substring_without_repeat.py
"""
Given a string s, find the length of the longest substring without repeating characters.

Example 1:
    Input: s = "abcabcbb"
    Output: 3
    Explanation: The answer is "abc", with the length of 3.

Example 2:
    Input: s = "bbbbb"
    Output: 1
    Explanation: The answer is "b", with the length of 1.

Constraints:
    0 <= s.length <= 5 * 10^4
    s (str): english letters, digits, symbols, and spaces
"""
#import
import unittest
#solution
class Solution(unittest.TestCase):
    def self_sol(self, s: str) -> int:
        """
        Given a string s, find the length of the longest substring without repeating characters.


        Pseudocode:
            if |string| in {0,1} 
                return |string|
            else
                string_size <- 0
                string_start <- 0
                string_end <- 0
                while string_end < |string|
                    if s[string_end] is not in string from start to end then
                        string_end <- n + 1
                        string_size <- max of string_size and (string_end - string_start)
                    else
                        string_start <- string_start + 1
                return string_size

        time complexity -> O(n)
        space complexity -> O(1)

        Args:
            s (str): input string

        Returns:
            int: length of longest substring without repeating characters
        """
        if len(s) in (0,1):
            return len(s)
        
        length = 0
        i = 0
        n = 0
        while n < len(s):
            if s[n] not in s[i:n]:
                n += 1
                length = max(length, n - i)
            else:
                i += 1
        return length
    
    def test_1(self):
        self.assertEqual(self.lengthOfLongestSubstring("abcabcbb"), 3)

    def test_2(self):
        self.assertEqual(self.lengthOfLongestSubstring("bbbbb"), 1)

    def test_3(self):
        self.assertEqual(self.lengthOfLongestSubstring("pwwkew"), 3)

if __name__ == "__main__":
    unittest.main()
    
    

File: algorithms/leetcode/longest-substring-without-repeating-characters/longest_substring_without_repeat.cpp
/*
SF

Given a string, find the length of the longest substring without repeating characters.

Example 1:
Input: "abcabcbb" 
Output: 3

Example 2:
Input: "bbbbb"
Output: 1

Example 3:
Input: "pwwkew"
Output: 3
*/
//imports
#include <iostream>
#include <string>
#include <vector>
#include <unordered_map>
#include <unordered_set>
#include <algorithm>

using namespace std;
//data structure
class Solution {
public:
    //all unique
    bool all_unique(string str, int start, int end) {
        unordered_set<char> set;
        for (int i = start; i < end; i++) {
            char ch = str[i];
            if (set.find(ch) != set.end()) return false;
            set.insert(ch);
        }
        return true;
    }
    //brute force
    int brute_force_solution(string str) {
        int n = str.length(), ans{0};
        
        for (auto i{0}; i < n; i++) {
            for (auto j{i + 1}; j <= n; j++) {
                if (all_unique(str, i, j)) ans = max(ans, j - i);
            }
        }
        return ans;
    }
    //map
    int map_solution(string str) {
        //find the length of the longest substring without repeating characters
        int n = str.length(), ans{0};
        //unordered_map to store the index of the last occurrence of each character
        unordered_map<char, int> map;
        //two pointers, i and j
        for (auto j{0}, i{0}; j < n; j++) {
            //if the character is already in the map, move the pointer i to the right of the last occurrence of the character
            if (map.find(str[j]) != map.end()) i = max(map[str[j]], i);
            //update the answer
            ans = max(ans, j - i + 1);
            //update the index of the last occurrence of the character
            map[str[j]] = j + 1;
        }
        return ans;
    }
};

//run
int main(int argc, char **argv) {
    Solution s;
    auto strings = vector<string>();
    if (argc < 2) {
        string strs[] = {"abcabcbb", "bbbbb", "pwwkew"};

        for (string str : strs) strings.push_back(str);
        
    } else {
        for (auto i{1}; i < argc; i++) strings.push_back(argv[i]);
    }

    auto header{"(brute_force_solution, map_solution) ->"};
    for (auto str : strings) {
        auto first{str == strings.at(0)};
        if (first) cout << header;

        cout << "\t(" << s.brute_force_solution(str) << ',' 
            << s.map_solution(str) << ')';
    }
    return 0;
}